### PR TITLE
feat: 2FA enforcement - database

### PR DESCRIPTION
**Pull request by WcaleNieWolny**
### Changes (AI generated)

- **Section 1: has_2fa_enabled functions**
  - Added `has_2fa_enabled()` function that checks if the current authenticated
    user has any verified MFA factors
  - Added `has_2fa_enabled(user_id uuid)` function that checks if a specific
    user has any verified MFA factors
  - The no-argument version is accessible to `authenticated` and `anon` roles
  - The `user_id` version is private and only accessible to `postgres` and
    `service_role` roles
  - Both functions use `SECURITY DEFINER` to access `auth.mfa_factors` table

- **Section 2: check_org_members_2fa_enabled function**
  - Added `check_org_members_2fa_enabled(org_id uuid)` function that returns a
    list of all organization members with their 2FA status
  - Function is accessible only to `super_admins` of the organization
  - Returns a table with `user_id` and `2fa_enabled` columns for all members
  - Validates that the organization exists and that the caller has super_admin
    rights

- **Section 3: Add enforcing_2fa column to orgs table**
  - Added `enforcing_2fa` boolean column to `public.orgs` table with
    `NOT NULL DEFAULT false`
  - Column indicates whether an organization requires all members to have 2FA
    enabled
  - Added documentation comment explaining the column's purpose

- **Section 4: Modify check_min_rights to enforce 2FA**
  - Updated `check_min_rights(user_id uuid, org_id uuid, ...)` function to check
    2FA enforcement rules
  - If an organization has `enforcing_2fa` enabled and the user doesn't have 2FA
    enabled, the function returns `false` and logs the denial
  - The 2FA check happens before permission checks, ensuring users without 2FA
    are denied access even if they would otherwise have permissions
  - Maintains backward compatibility - if org doesn't enforce 2FA, behavior is
    unchanged

- **Section 4.A: Create get_orgs_v7 with 2FA enforcement**
  - Created new `get_orgs_v7(userid uuid)` function based on `get_orgs_v6` with
    2FA enforcement support
  - Added `enforcing_2fa` field to the return table showing whether the org
    requires 2FA
  - Added `2fa_has_access` field indicating if the user has access (true if org
    doesn't enforce 2FA OR user has 2FA enabled)
  - Redacts sensitive fields (`paying`, `trial_left`, `can_use_more`,
    `is_canceled`, `app_count`, `subscription_start`, `subscription_end`,
    `management_email`, `is_yearly`) when user doesn't have 2FA access
  - Uses a CTE (`two_fa_access`) to compute 2FA access status efficiently
  - Created wrapper function `get_orgs_v7()` that handles API key and
    authentication logic

- **Section 4.B: Modify get_orgs_v6 to prevent information leakage**
  - Updated `get_orgs_v6(userid uuid)` function to redact sensitive information
    when users don't have 2FA access
  - Prevents information leakage by returning redacted values (false, 0, NULL)
    for sensitive fields when org enforces 2FA and user doesn't have it
  - Uses the same `two_fa_access` CTE pattern as `get_orgs_v7` for consistency
  - Maintains backward compatibility - all existing fields remain, just with
    redaction logic added

- **Section 5: reject_access_due_to_2fa function**
  - Added `reject_access_due_to_2fa(org_id uuid, user_id uuid)` function that
    returns boolean
  - Returns `true` if org requires 2FA and user doesn't have it enabled, `false`
    otherwise
  - Returns `false` if org doesn't enforce 2FA enforcement
  - Returns `false` if org doesn't exist
  - Function is private and only accessible to `postgres` and `service_role`
    roles (not accessible to regular users)
  - Logs access denials using `pg_log` for audit purposes

- **Tests**
  - Added comprehensive SQL tests in `35_test_has_2fa_enabled.sql` covering both
    function variants and permission checks
  - Added comprehensive SQL tests in `36_test_check_org_members_2fa_enabled.sql`
    covering super_admin access, member 2FA status, and edge cases
  - Added extensive SQL tests in `37_test_check_min_rights_2fa_enforcement.sql`
    covering 2FA enforcement blocking access, normal access scenarios, different
    permission levels, and edge cases
  - Added comprehensive SQL tests in `38_test_get_orgs_v7_2fa_enforcement.sql`
    covering field visibility, redaction verification, and comparison between
    users with and without 2FA
  - Added SQL tests in `39_test_reject_access_due_to_2fa.sql` covering function
    behavior, permission checks, and verification that regular users cannot call
    the private function
  - All tests verify both positive and negative cases, edge cases, and
    permission boundaries

### Motivation

This PR is the direct resonse to part one of #1291. I strongly believe that the changes of this PR bring Capgo closer to fully implemeneting the 2FA enforcement feature. As such, I believe this PR is a high priority for the development of capgo and it should be reviewed and merged as soon as possible, so further development of the 2FA enforcement can continue.

### Business impact

The feature of 2FA enforcement has been deemed important. (Medium business impact. Enterprise clients often consider 2FA enforcement to provide good operation security for their organization). However, this PR on its own does not fully implement the 2FA enforcement feature, so on its own this PR doesn't have a business impact.

### Questions to be ansered during the review

**Q: Should the get_orgs_v6 be modified to prevent the leak of sensitive data when 2FA enforcement is enabled?**

~~**Q: How to lint the SQL?**~~

I ran `npm run lint:sql` in 278e0d1b1719368c05ee9f602ee5ebb1f85d8437

## Summary by CodeRabbit

* **New Features**
  * Organization admins can require two-factor authentication for members.
  * Users without 2FA are blocked from accessing orgs that enforce it.
  * Org listings surface an "enforcing 2FA" flag and per-user 2FA access; sensitive org fields are redacted when 2FA access is missing.

* **Tests**
  * Added comprehensive tests covering 2FA status checks, enforcement behavior, access control permutations, redaction logic, and permission boundaries.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

---

> [!NOTE]
> Implements database-side 2FA enforcement and redaction to protect org data when users lack verified MFA.
> 
> - Add `enforcing_2fa` column to `orgs`; update `check_min_rights` to deny access when enforced and `has_2fa_enabled(user)` is false (logged via `pg_log`).
> - New functions: `has_2fa_enabled()`/`has_2fa_enabled(uuid)` (MFA status), `check_org_members_2fa_enabled(org_id)` (super_admin-only member 2FA status), `reject_access_due_to_2fa(org_id, user_id)` (backend-only gating helper).
> - Add `get_orgs_v7(userid)` and wrapper `get_orgs_v7()` returning `enforcing_2fa` and `2fa_has_access`; redact sensitive fields when access is blocked. Modify `get_orgs_v6(userid)` to apply equivalent redaction.
> - Tighten grants: user-scoped helpers public where safe; `uuid` variants and backend helpers restricted to `postgres/service_role`.
> - Add extensive SQL tests covering MFA detection, org-member checks, rights enforcement across scopes, v6/v7 redaction behavior, and function permissions.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 287fd7ccf7e55c6a0fbbe35fb8c068bcf8fb9123. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: supabase/migrations/20251224103713_2fa_enforcement.sql [hunk 1]
```diff
@@ -0,0 +1,600 @@
...
+  JOIN public.org_users ou ON ou.user_id = userid AND o.id = ou.org_id
+  JOIN two_fa_access tfa ON tfa.org_id = o.id
+  LEFT JOIN public.stripe_info si ON o.customer_id = si.customer_id
+  LEFT JOIN public.plans p ON si.product_id = p.stripe_id
+  LEFT JOIN app_counts ac ON ac.owner_org = o.id
+  LEFT JOIN public.usage_credit_balances ucb ON ucb.org_id = o.id
+  LEFT JOIN paying_orgs_ordered poo ON poo.id = o.id
+  LEFT JOIN billing_cycles bc ON bc.org_id = o.id;
+END;
+$$;
```
[the remark this answers is not in the document]
Comment (WcaleNieWolny) [reply supabase/migrations/20251224103713_2fa_enforcement.sql:544]: wait, this is bad no? orgs_v6(user_id) should only be callable by supabase/postgres and not anyone no? This new code is technically better? perhaps I should also do code to revoke from anon/authenticated?

[the remark this answers is not in the document]
Comment (WcaleNieWolny) [reply supabase/migrations/20251224103713_2fa_enforcement.sql:544]: Fixed in b6dc5a0dd3a3208b85a1e60615af2ea15454acf6, can you check?

File: supabase/migrations/20251224103713_2fa_enforcement.sql [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+        public.check_min_rights(
+            'super_admin'::public.user_min_right,
+            (SELECT public.get_identity_org_allowed('{read,upload,write,all}'::public.key_mode[], check_org_members_2fa_enabled.org_id)),
+            check_org_members_2fa_enabled.org_id,
+            NULL::character varying,
+            NULL::bigint
+        )
+    ) THEN
+        RAISE EXCEPTION 'NO_RIGHTS';
+    END IF;
```
[the remark this answers is not in the document]
Comment (WcaleNieWolny) [reply supabase/migrations/20251224103713_2fa_enforcement.sql:96]: This makes no sense. It's exactly what I want. If a super_admin doesn't have 2fa enabled and the org has 2fa enforcement on, the super admin SHOULD be denied acess to this fn
